### DISCUSSION TITLE
Reading nested fields from environment variables

### DISCUSSION THREAD

**Discussion by m2m1** [pydantic/pydantic#4517] [Question]
I use Pydantic to read environment variables. In principle it works and the variables can be read and stored. There is a mechanism to specify a name for an environment variable that should be read for a field `var: int = Field(..., env='desired_name')`. For a top level `BaseSettings` this works expected and the value of `DESIRED_NAME` is stored in `var`.

However if I use nested structures the `env` keyword for a field seems to have no effect at all and the environment variable isn't read. In #4515 I posted a sample snippet. @hramezani Thank you for your fast answer but I don't understand it properly.

When I read about the `env` keyword I expected that regardless of the model and the nesting structure with `env=` I can specify a name for the environment variable that should be used for this Field. Apparently this is not the case. As I understand your answer it implies that I can only specify the name for environment variables for top level models, right? Or is there an other way to achieve my goal?

**Comment by samuelcolvin** [comment]
See [the docs on "Parsing environment variable values"](https://pydantic-docs.helpmanual.io/usage/settings/#parsing-environment-variable-values).

If you want to fill the nested field, use `sub__sub_var` (note the double underscore to identify the nested field).

**Reply by m2m1** [reply]
A colleague proposed `sub__desired_name` should be read for sub_var but IMHO this defeats the purpose in the first place.

It seemed odd to me that `env=` doesn't have the same effect for fields and for nested fields.

Would you mind to clarify in the docs that `env=` only works for non-nested fields?

I can also open a feature request if this has some chance to be implemented.

**Reply by samuelcolvin** [reply]
settings is moving to the new [pydantic-settings](https://github.com/pydantic/pydantic-settings) package for V2. We're going to make env var. parsing logic a "source" in V2 so it can be customised and replaced much more easily.

Feel free to create a feature request issue on the new repo.

**Reply by m2m1** [reply]
Alright, that makes sense.

Slightly different, but similar question: Can I read secrets for nested fields? What is the syntax for it? `t_var` for a field did, `sub__sub_var` for a nested field didn't work.

**Reply by samuelcolvin** [reply]
not sure, best to try it or read the code.
